### ISSUE TITLE
zrepl replicates every snapshot including those created outside zrepl

### ISSUE THREAD

**Issue by calistoc** [zrepl/zrepl#403]
Hello,
just to ask some help please as I was searching the documentation but I’m confused and not sure whether I’m missing something.

For what I see zrepl replicates all the snapshots in a filesystem including those snaps created outside zrepl, for example I have some snaps created to perform a daily backup that I would like not to be considered by zrepl but they get replicated once I wakeup the jobs.
This happens for push jobs with periodic snapshotting and also with jobs with manual snapshotting where I need to wakeup the jobs via crontab and where I cannot specify a prefix in the snapshotting configuration:

snapshotting:
    type: manual
    prefix: zrepl_ (it gives this error:  field prefix not found in type config.SnapshottingManual)

It seems to me that the snapshotting prefix for push jobs with periodic snapshotting is only used for pruning but it is not used to select the snapshots to be replicated, is this correct?
Is there a way to have zrepl replicating only its own snapshots or better only those specified by a prefix?

Thank you

**Comment by problame** [comment]
Hi, thanks for describing your use case.

> It seems to me that the snapshotting prefix for push jobs with periodic snapshotting is only used for pruning but it is not used to select the snapshots to be replicated, is this correct?

Actually, it's even less than that. The prefix you specify in the `snapshotting` section is really only used to define the snapshot name, nothing more, nothing less.

>  Is there a way to have zrepl replicating only its own snapshots or better only those specified by a prefix?

No, currently there isn't a way to tell zrepl to do that.

- #248 requests prefix-based filtering of snapshots for replication. It seems that if we implemented prefix-based filtering, your problem would be solved. Wouldn't it?
- It seems like you assumed zrepl would somehow 'own' the snapshots it created and either only act on the snapshots it created or treat them specially during replication or pruning. That is currently not the case - replication and pruning do not know whether a snapshot was created by zrepl or something else. But it's a very widespread belief, and zrepl should probably develop into a direction where snapshot management and replication are more tightly integrated.

If you feel like #248 would be the idea solution, please subscribe there and post a proposal for the configuration syntax in your scenario.

**Comment by calistoc** [comment]
Thanks for your feedbacks

Yes I think the [#248](https://github.com/zrepl/zrepl/issues/248) request would solve my problem.
About the proposal for the configuration syntax, I would add the following optional parameter in the configuration job, either push, pull or source, something like this:

- type: push
  name: to_server2
  replication_snapshots_prefix: zrepl_
…

This would replicate only snapshots with zrepl_ prefix, not specifying this parameter would leave things as they are now replicating all snapshots and preserving the current functionalities
I posted the same on [#248](https://github.com/zrepl/zrepl/issues/248)

Thank you for your work!

**Comment by beren12** [comment]
Just a few things. If you send a range of snapshots for incremental sends, then every snap in between would be sent, no way to not do that unless you broke it up into smaller chunks. You could manually destroy or prune non-zrepl snaps on the remote side. I believe you can set custom properties on snapshots, there was a patch for zfs-auto-snap to do this to mark things like daily/weekly/monthly snapshots while all having the same name prefix. zrepl could set some properties and look for it for filtering which snaps to send/prune

**Comment by problame** [comment]
> Just a few things. If you send a range of snapshots for incremental sends, then every snap in between would be sent, no way to not do that unless you broke it up into smaller chunks

@beren12  What are you referring to here? {desired, current} zrepl behavior or `zfs send -I`?

> You could manually destroy or prune non-zrepl snaps on the remote side. 

What do you mean by 'non-zrepl snaps'? Snapshtos not created by zrepl on the sender but replicated by zrepl to the receiver?

> I believe you can set custom properties on snapshots, there was a patch for zfs-auto-snap to do this to mark things like daily/weekly/monthly snapshots while all having the same name prefix. zrepl could set some properties and look for it for filtering which snaps to send/prune

There are two problems with attaching policy to snapshots via snapshot user properties:
- The first user property on a snapshot allocates the ZAP object for it and thus has some storage overhead. Probably negligible.
- If property replication is enabled, the snapshot property that was `zfs set` on the sending system is also present on the receiving system.
  - If it's a non-zrepl policy that is a problem the user has to deal with.
  - If it's a zrepl policy I'd want zrepl to be smart about it.
    - What should zrepl do with the snapshot property that would make a snapshot eligible for sending even if it's not matched by the to-be-done `replication_snapshots_prefix`?
    - What should zrepl do if a snapshot is marked as a `ghost` for pruning (i.e. pruning shouldn't even consider it)?

**Comment by problame** [comment]
(I originally marked this issue as a duplicate of #248 but it's actually the first original request for a 'filtering' kind of feature).

**Comment by beren12** [comment]
Sorry what I was trying to say was if you send incremental sends, there's no way to avoid this unless you prune the snapshot before, which some people don't want, or after it's send prune from the remote side. With the properties… I didn't think too much of the downsides to deal with, just threw that idea out there. But yes, you would need to have some way to tag the properties per host, like with the hostname but even then that may not be unique.

**Comment by beren12** [comment]
I guess you could stop early if the latest snapshot(s) weren't from zrepl though

**Comment by problame** [comment]
>  Sorry what I was trying to say was if you send incremental sends, there's no way to avoid this unless you prune the snapshot before, which some people don't want, or after it's send prune from the remote side. 

What do you mean by *this* in this sentence? Please inline-quote what are referring to.

> I guess you could stop early if the latest snapshot(s) weren't from zrepl though

zrepl doesn't have a concept of 'ownership' or 'knowing which snapshots it created' at this time. ([see above](https://github.com/zrepl/zrepl/issues/403#issuecomment-750198008)). I don't understand what you mean by 'stop early'.

**Comment by problame** [comment]
(Another remark to this feature request in general)

In  https://github.com/zrepl/zrepl/issues/248#issuecomment-559726992 I proposed to implement snapshot filtering as a sender-side feature, i.e., only snapshots matching the filter would go over the wire. I anticipated the following problem:

> The functionality was integrated such that the replication logic was only presented the filtered view of the sender's snapshots.
Now, when a sender changed their filter to include manually taken snapshots, the replication logic would be ... confused ... because suddenly there were snapshots on the sender that weren't there before, and that were older than the last successful replication.

We can work around this in two ways:
- Do the snapshot filtering in the replication code, not the endpoint code
  - => Only downside is trust implications: sender has to trust receiver with all snapshots on a filesystem.
    - Can't think of a use case where this is relevant. Can anyone?
- Filter in the endpoint code but always include the replication cursor bookmark and/or the most recent step hold snapshot(s).

With both solutions we'd rely on the diffing code to not look at snapshots older than the most recent common ancestor, but that's the case already (needs test coverage!).

Note about the case where the user changes the snapshot filter so that a resuming step would no longer be done at all if it was an non-resuming step: we could also abort the resume and use the step bookmark as incremental source + whatever next snapshot matches the snapshot filter. That's probably what the user wants anyways.
- We can inspect the resume token

**Comment by problame** [comment]
@InsanePrawn thoughts on [this](https://github.com/zrepl/zrepl/issues/403#issuecomment-784548672)?

**Comment by InsanePrawn** [comment]
> In [#248 (comment)](https://github.com/zrepl/zrepl/issues/248#issuecomment-559726992) I proposed to implement snapshot filtering as a sender-side feature, i.e., only snapshots matching the filter would go over the wire. I anticipated the following problem:
> 
> > The functionality was integrated such that the replication logic was only presented the filtered view of the sender's snapshots.
> > Now, when a sender changed their filter to include manually taken snapshots, the replication logic would be ... confused ... because suddenly there were snapshots on the sender that weren't there before, and that were older than the last successful replication.

Sounds like the situation is not to get confused:

> With both solutions we'd rely on the diffing code to not look at snapshots older than the most recent common ancestor, but that's the case already (needs test coverage!).

Exactly. I think this happens indirectly already when you prune your sides at different rates, i.e. with different grids?

> Note about the case where the user changes the snapshot filter so that a resuming step would no longer be done at all if it was an non-resuming step: we could also abort the resume and use the step bookmark as incremental source + whatever next snapshot matches the snapshot filter. That's probably what the user wants anyways.

Can't we try to repl, notice we have a newer 'unknown' snapshot, ask the other side if it also has that (circumvents the filter here!) and then see if we can do mitigation that way? (If the other side says no and future conflict resolution strategy `destroy` is selected, we roll back to the common ancestor and resync)
That way a malicious remote can not exfiltrate otherwise filtered snapshots but only get incremental streams that start _on_ otherwise filtered snapshot, meaning you don't get the filtered snapshot itself ever.
Tl;dr: Do more communication as fallback/if it becomes evident during planning, make sure to allow filtered snaps to be used as `From`, never `To`. That should work, right?

**Comment by calistoc** [comment]
Not sure this will address all the above considerations but I had some time and I tried to figure out how to implement the snapshots prefix filtering using the following syntax:

-  type: push
   replication_snapshots_prefix: “zrepl_”
   …

The result was https://github.com/zrepl/zrepl/pull/459 that adds the optional parameter replication_snapshots_prefix for push and source jobs, if not used zrepl will keep working as it used to.
I have been testing it for a few days and so far it is working fine, it needs more testing anyway.

Please let me know what you think about it

**Comment by problame** [comment]
@calistoc sorry this took so long.

---

In comment https://github.com/zrepl/zrepl/issues/403#issuecomment-784548672 I presented the following two options for an implementation:

> 
> * Do the snapshot filtering in the replication code, not the endpoint code
>    *  => Only downside is trust implications: sender has to trust receiver with all snapshots on a filesystem.
>             Can't think of a use case where this is relevant. Can anyone?
> * Filter in the endpoint code but always include the replication cursor bookmark and/or the most recent step hold snapshot(s).

#459 implements the latter.
I'm still undecided which option is better from a UX / long-term maintenenance perspective but your PR solves a specific issue that some people have with zrepl right now: they only want zrepl to replicate the snapshot that it creates itself, the underlying assumption being that the replication code knows what snapshots the snapshotter created (it doesn't).

Some specific concerns:
- the hard-coded `zrepl_CURSOR` can be relaxed. As mentioned previously in this thread I think we can just present all bookmarks. Maybe `ShortnamePrefix` should be renamed to `SnapShortnamePrefix`.
- What about people who want to filter multiple prefixes? I'm thinking about a set of cronjobs that creates `hourly_...`, `daily_...`, `weekly_...` and `monthly...` snapshots where the user wants to specify that only `daily`, `weekly` and `monthly` should be replicated. A regex could do, but maybe a list would be more expressive?
  - I'm not asking you to implement this right away, just want to know what people subscribed to this issue think.
- Have you played around with what happens if you change the value of `replication_snapshots_prefix` in the following situations:
  - Interrupt a running resumable replication, change the config to exclude the snapshots in the replication step, restart. What would you expect to happen in that case?
  - Test that you can't provoke a divergence:
    - Create send-side snaps `@A1 @B1 @A2`, replicating with allowed prefix `A` (`@A1, @A2`), then change config to allow all prefixes and trigger replication again. Should _not_ complain about `@B1` and just tell you that everything is replicated.

**Comment by calistoc** [comment]
> 
>     * Have you played around with what happens if you change the value of `replication_snapshots_prefix` in the following situations:
>       
>       * Interrupt a running resumable replication, change the config to exclude the snapshots in the replication step, restart. What would you expect to happen in that case?

It will fail to resume with error:

bck/test PLANNING-ERROR (step 0/0, 0 B/0 B) resume token `toguid` = 13902766327137181066 not found on sender (`toname` = "bck/test@cc1")

I think this would be expected anyway, imho this kind of changes should be done rarely and paying attention about what is being done, that is stopping zrepl, making sure the snapshots replication is completed, change the configuration and then restart zrepl.
I think that a note could be added in the documentation though.

>       * Test that you can't provoke a divergence:
>         
>         * Create send-side snaps `@A1 @B1 @A2`, replicating with allowed prefix `A` (`@A1, @A2`), then change config to allow all prefixes and trigger replication again. Should _not_ complain about `@B1` and just tell you that everything is replicated.

Yes that is correct, zrepl will tell all is replicated not complaining about `@B1`

**Comment by problame** [comment]
> It will fail to resume with error:
> 
> bck/test PLANNING-ERROR (step 0/0, 0 B/0 B) resume token toguid = 13902766327137181066 not found on sender (toname = "bck/test@cc1")
> 
> I think this would be expected anyway, imho this kind of changes should be done rarely and paying attention about what is being done, that is stopping zrepl making sure the snapshots replication is completed, change the configuration and then restart zrepl.
> I think that a note could be added in the documentation though.

Yeah, I wish we could make this smoother but I don't see it happening for the foreseeable future.

> Yes that is correct, zrepl will tell all is replicated not complaining about @B1

Great.

> #459 implements the latter.
I'm still undecided which option is better from a UX / long-term maintenenance perspective but your PR solves a specific issue that some people have with zrepl right now: they only want zrepl to replicate the snapshot that it creates itself, the underlying assumption being that the replication code knows what snapshots the snapshotter created (it doesn't).

I'll have to think about the pros and cons of doing the filtering on the sender instead of on the active side (active side = where the replication logic runs). I'm interested in your input, also cc'ing @InsanePrawn , maybe @grahamc ?

**Comment by psy0rz** [comment]
ZFS-autobackup does this correctly and has an option to include other snapshots as well.

Please check it out at: https://github.com/psy0rz/zfs_autobackup

(I suggest the latest beta of 3.1, its reliable)

**Comment by InsanePrawn** [comment]
> ZFS-autobackup does this correctly and has an option to include other snapshots as well.
> 
> Please check it out at: 

weird advertisement strat but ok

**Comment by antifuchs** [comment]
I just ran into a problem that's kind of related to this one: I have both [zfs-auto-snapshot](https://github.com/bdrewery/zfstools) and zrepl running (they both serve a purpose, it's annoying to explain so please believe me that they are both necessary at the moment), and zrepl transmits the zfs-auto-snapshot snapshots to the destination, where the pruning policy promptly tries to delete them. But the pruning policy fails because on occasion, the "frequent" job runs fast enough that zrepl apparently considers the last frequent snapshot the most recent one and puts a bookmark or hold on it, which then regularly results in errors in backup pruning.

Example log line from the prune failure:
```console
Nov 19 13:39:19 gloria zrepl[3731041]: 2024-11-19T13:39:19-05:00 [ERRO][gloria][pruning][wonX$Ejwd$Ejwd.ihex.PR1W.YDqb]: target could not destroy snapshots prune_side="receiver" err="destroys failed @zfs-auto-snap_frequent-2024-11-19-13h30: dataset is busy"
```

My pruning policy is as follows:
```yaml
  pruning:
    keep_receiver:
    - grid: 1x1h(keep=all) | 24x1h | 30x1d | 6x30d
      regex: ^zrepl_
      type: grid
    keep_sender:
    - type: not_replicated
    - negate: true
      regex: ^zrepl_
      type: regex
```

On the surface, it seems like a pretty easy optimization - don't transfer snapshots that would get pruned on the receiver side anyway (:
